debian/aa.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH AA 1 "September 4, 2006"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
aa \- astronomical almanac - calculate planet and star positions
.SH SYNOPSIS
.B aa
.SH DESCRIPTION
The \fBaa\fP program computes the orbital positions of planetary
bodies and performs rigorous coordinate reductions to apparent
geocentric and topocentric place (local altitude and azimuth).  It
also reduces star catalogue positions given in either the FK4 or FK5
system.  Data for the 57 navigational stars is included.  Most of the
algorithms employed are from The Astronomical Almanac (AA) published
by the U.S. Government Printing Office.

The \fBaa\fP program follows the rigorous algorithms for reduction of
celestial coordinates exactly as laid out in current editions of the
Astronomical Almanac.  The reduction to apparent geocentric place has
been checked by a special version of the program (aa200) that takes
planetary positions directly from the Jet Propulsion Laboratory DE200
numerical integration of the solar system. The results agree exactly
with the Astronomical Almanac tables from 1987 onward (earlier
Almanacs used slightly different reduction methods).
.SH Initialization
The following items will be read in automatically from the first of
these files to be found: \fI./aa.ini\fP, \fI~/.aa.ini\fP,
\fI/etc/aa.ini\fP.  The file contains one ASCII string number per line
so is easily edited.  A sample initialization file is supplied.  The
entries are:

.IP lon
Terrestrial longitude of observer, degrees East of Greenwich
.IP lat
Geodetic latitude of observer (program calculates geocentric latitude)
.IP height
Height above sea level, meters
.IP temp
Atmospheric temperature, degrees Centigrade
.IP pressure
Atmospheric pressure, millibars
.IP tflag
Input time type: 1 = TDT, 2 = UT, 0 = TDT set equal to UT
.IP deltaT
Value to use for deltaT, seconds; if 0 then the program will compute it.
.SH Orbit Computations
   Several methods of calculating the positions of the planets
have been provided for in the program source code.  These range
in accuracy from a built-in computation using perturbation formulae
to a solution from precise orbital elements that you supply from
an almanac.
   The program uses as a default a set of trigonometric
expansions for the position of the Earth and planets.  These have
been adjusted to match the Jet Propulsion Laboratory's DE404 Long
Ephemeris (1995) with a precision ranging from about 0.1" for the
Earth to 1" for Pluto. The adjustment was carried out on the
interval from 3000 B.C. to 3000 A.D. for the outer planets.  The
adjustment for the inner planets is strictly valid only from 1350
B.C. to 3000 A.D., but may be used to 3000 B.C. with some loss of
precision.  See \fI/usr/share/doc/aa/readme.404\fP for additional information.  The true
accuracy of positions calculated for prehistoric or future dates
is of course unknown.
   The Moon's position is calculated by a modified version of the
lunar theory of Chapront-Touze' and Chapront.  This has a precision
of 0.5 arc second relative to DE404 for all dates between
1369 B.C. and 3000 A.D.  The real position of the Moon in ancient
times is not actually known this accurately, due to uncertainty
in the tidal acceleration of the Moon's orbit.

   In the absence of an interpolated polynomial ephemeris such
as the DE200, the highest accuracy for current planetary
positions is achieved by using the heliocentric orbital elements
that are published in the Astronomical Almanac. If precise
orbital elements are provided for the desired epoch then the
apparent place should be found to agree very closely with
Almanac tabulations.
   Entering 99 for the planet number generates a prompt for the
name of a file containing human-readable ASCII strings specifying
the elements of orbits. The items in the specification are
(see also the example file orbit.cat):

.RS
.nf
   First line of entry:
epoch of orbital elements (Julian date)
inclination
longitude of the ascending node
argument of the perihelion
mean distance (semimajor axis) in au
daily motion
.fi

.nf
   Second line of entry:
eccentricity
mean anomaly
epoch of equinox and ecliptic, Julian date
visual magnitude B(1,0) at 1au from earth and sun
equatorial semidiameter at 1au, arc seconds
name of the object, up to 15 characters
.fi
.RE

Angles in the above are in degrees except as noted.  Several
sample orbits are supplied in the file orbit.cat.  If you read
in an orbit named "Earth" the program will install the Earth
orbit, then loop back and ask for an orbit number again.
  The entry for daily motion is optional.  It will be calculated
by the program if it is set equal to 0.0 in your catalogue.
Almanac values of daily motion recognize the nonzero mass of the
orbiting planet; the program's calculation will assume the mass
is zero.
  Mean distance, for an elliptical orbit, is the length of the
semi-major axis of the ellipse. If the eccentricity is given to
be 1.0, the orbit is parabolic and the "mean distance" item is
taken to be the perihelion distance.  Similarly a hyperbolic
orbit has eccentricity > 1.0 and "mean distance" is again
interpreted to mean perihelion distance.  In both these cases,
the "epoch" is the perihelion date, and the mean anomaly is
set to 0.0 in your catalogue.
  Elliptical cometary orbits are usually catalogued in terms of
perihelion distance also, but you must convert this to mean
distance to be understood by the program. Use the formula

  mean distance = perihelion distance / (1 - eccentricity)

to calculate the value to be entered in your catalogue for an
elliptical orbit.
  The epoch of the orbital elements refers particularly to the
date to which the given mean anomaly applies.  Published data
for comets often give the time of perihelion passage as a
calendar date and fraction of a day in Ephemeris Time.  To
translate this into a Julian date for your catalogue entry, run
\fBaa\fP, type in the published date and decimal fraction of a
day, and note the displayed Julian date. This is the correct
Julian Ephemeris Date of the epoch for your catalogue entry.
Example (Sky & Telescope, March 1991, page 297): Comet Levy
1990c had a perihelion date given as 1990 Oct 24.68664 ET.  As
you are prompted separately for the year, month, and day, enter
1990, 10, 24.68664 into the program. This date and fraction
translates to JED 2448189.18664.  For comparison purposes, note
that published ephemerides for comets usually give astrometric
positions, not apparent positions.
.SH Ephemeris Time and Other Time Scales
   Exercise care about time scales when comparing results against
an almanac.  The orbit program assumes input date is Ephemeris
Time (ET or TDT).  Topocentric altitude and azimuth are
calculated from Universal Time (UT).  The program converts
between the two as required, but you must indicate whether your
input entry is TDT or UT.  This is done by the entry for input
time type in aa.ini.  If you are comparing positions against
almanac values, you probably want TDT.  If you are looking up at
the sky, you probably want UT.  Ephemeris transit times can be
obtained by declaring TDT = UT.  The adjustment for deltaT = ET
minus UT is accurate for the years 1620 through 2011, as the
complete tabulation from the Astronomical Almanac is included in
the program. Outside this range of years, approximate formulas
are used to estimate deltaT.  These formulas are based on
analyses of eclipse records going back to ancient times
(Stephenson and Houlden, 1986; Borkowski, 1988) but they do not
predict future values very accurately.  For precise calculations,
you should update the table in deltat.c from the current year's
Almanac. Note the civil time of day is UTC, which is adjusted by
integral leap seconds to be within 0.9 second of UT.

   Updated deltaT values and predictions can be obtained from this
network archive: http://maia.usno.navy.mil .
See the file deltat.c for additional information.
In addition, the IAU has adopted several other definitions of
time, but this program does not distinguish among them.  The
International Earth Rotation Service is in charge of UT. Precise
data on Earth rotation and orientation are published in the IERS
bulletins, available at the IERS computer site www.iers.org as well
as at the usno site.
.SH Rise and Set Times
   Each calculation of the time of local rising, meridian transit, and
setting includes a first order correction for the motion in right
ascension and declination of the object between the entered input time
and the time of the event.  Even so, the calculation has to be
iterated, or repeated with successively closer estimates of the event
time.  In view of the first order correction the iteration has a
second-order convergence characteristic and arrives at a precise
result in just two or three steps.  On the other hand, the technique
used is unstable for nearly-circumpolar objects, such as the Moon
observed at high latitudes.  Thus a failure to report rise and set
times does not necessarily mean that there was no rise or set event.

   The program reports the transit that is nearest to the input time.
Rise and set times ordinarily precede and follow the transit.  Check
the date displayed next to the rise, set, or transit time to be sure
the results are for the desired date and not for the previous or next
calendar day.  For the Sun and Moon, rise and set times are for the
upper limb of the disc; but the indicated topocentric altitude always
refers to the center of the disc.  The computed event times include
the effects of diurnal aberration and parallax.

   Age of the Moon, in days from the nearest Quarter, also has a
correction for orbital motion, but does not get the benefit of
iterative improvement and may be off by 0.1 day (the stated Quarter is
always correct, however). The estimated time can be made much more
precise by entering the input date and time of day to be near the time
of the event.  In other words, the rigorous calculation requires
iterating on the time; in this case the program does not do so
automatically, hence if you want maximum accuracy you must do the
iteration by hand.
.SH Stars
   Positions and proper motions of the 57 navigational stars
were taken from the Fifth Fundamental Catalogue (FK5). They are
in the file \fI/usr/share/aa/star.cat\fP.  For all of these, the program's output of
astrometric position agreed with the 1986 AA to the precision of
the AA tabulation (an arc second).  The same is true for 1950
FK4 positions taken from the SAO catalogue.  The program agrees
to 0.01" with worked examples presented in the AA. Spot checks
against Apparent Places of Fundamental Stars confirm the mean
place agreement to <0.1".  The APFS uses an older nutation
series, so direct comparison of apparent place is difficult. 
The program incorporates the complete IAU Theory of Nutation
(1980).  Items for the Messier catalogue, \fI/usr/share/aa/messier.cat\fP, are from
either the AA or Sky Catalogue 2000.
   To compute a star's apparent position, its motion since the
catalogue epoch is taken into account as well as the
changes due to precession of the equatorial coordinate system.
Star catalogue files have the following data structure.  Each
star entry occupies one line of ASCII characters.  Numbers can
be in any usual decimal computer format and are separated from
each other by one or more spaces. From the beginning of the
line, the parameters are

.RS
.nf
Epoch of catalogue coordinates and equinox
Right ascension, hours
Right ascension, minutes
Right ascension, seconds
Declination, degrees
Declination, minutes
Declination, seconds
Proper motion in R.A., s/century
Proper motion in Dec., "/century
Radial velocity, km/s
Distance, parsecs
Visual magnitude
Object name
.fi
.RE
For example, the line

2000 02 31 48.704  89 15 50.72 19.877 -1.52 -17.0 0.0070 2.02 alUMi(Polaris)

has the following interpretation:

.RS
.nf
J2000.0      ;Epoch of coordinates, equator, and equinox
2h 31m 48.704s    ;Right Ascension
89deg 15' 50.72"   ;Declination
19.877       ;proper motion in R.A., s/century
-1.52        ;proper motion in Dec., "/century
-17.0        ;radial velocity, km/s
0.007        ;parallax, "
2.02         ;magnitude
alUMi(Polaris)    ;abbreviated name for alpha Ursae Minoris (Polaris)
.fi
.RE

   Standard abbreviations for 88 constellation names are
expanded into spelled-out form (see constel.c). The program
accepts two types of catalogue coordinates.  If the epoch is
given as 1950, the entire entry is interpreted as an FK4 item. 
The program then automatically converts the data to the FK5
system.  All other epochs are interpreted as being in the FK5
system.
   Note that catalogue (and AA) star coordinates are referred to
the center of the solar system, whereas the program displays the
correct geocentric direction of the object.  The maximum
difference is 0.8" in the case of alpha Centauri.
.SH OPTIONS
\fBaa\fP does not accept any options.
.SH FILES
.IR "\fI./aa.ini\fP, \fI~/.aa.ini\fP, \fI/etc/aa.ini\fP"
Initialization data.
.IP \fI/usr/share/doc/aa/readme.404\fP
Documentation of plan404 ephemerides.
.IP \fI/usr/share/aa/star.cat\fP
Catalogue data on the 57 navigational stars.
.IP \fI/usr/share/aa/messier.cat\fP
Items for the Messier catalogue
.SH SEE ALSO
\fBconjunct(1)\fP
.br
.SH AUTHOR
aa was written by Stephen L. Moshier <steve@moshier.net>.
.PP
This manual page was written by James R. Van Zandt <jrv@debian.org>,
for the Debian project (but may be used by others).


debian/changelog
astronomical-almanac (5.6-3) unstable; urgency=low

  * debian/control: move homepage reference to Homepage field.  Bump
    policy version to 3.8.0.
  
  * debian/rules: update handling of DEB_BUILD_OPTIONS

 -- James R. Van Zandt <jrv@debian.org>  Thu, 14 Aug 2008 18:49:09 -0400

astronomical-almanac (5.6-2) unstable; urgency=low

  * rplanet.c: declare robject as static, so it doesn't conflict with the
    definition in aa (suggested by author Steven Moshier).

  * debian/manpage.1.ex: delete 

  * /usr/share/doc/astronomical-almanac/examples/README: mention sample
    celestial navigation problem.  README is now a shell script.
  
  * /usr/share/doc/astronomical-almanac/examples/: add solution of sample
    celestial navigation problem.

 -- James R. Van Zandt <jrv@debian.org>  Mon,  9 Oct 2006 21:35:39 -0400

astronomical-almanac (5.6-1) unstable; urgency=low

  * Initial release (Closes:Bug#386034)
  * kfiles.c: default locations for star.cat and orbit.cat are in
    /usr/share/aa. look for initialization in ./aa.ini then ~/.aa.ini then
    /etc/aa.ini.  Look for star charts in /usr/share/aa/star.cat.  Look
    for orbits in /usr/share/aa/orbit.cat.
  * conjunct.c: Accept inputs from the command line (start and end time,
    event type).
  * aa.que, aa.ans:   Look for star charts in /usr/share/aa/star.cat.  Look
    for orbits in /usr/share/aa/orbit.cat.
  * dms.c: check whether fgets returns NULL.

 -- James R. Van Zandt <jrv@debian.org>  Mon,  4 Sep 2006 15:13:08 -0400



aa.c
/* This program calculates orbits of planetary bodies and reduces
 * the coordinates of planets or stars to geocentric and topocentric
 * place.  An effort has been made to use rigorous methods throughout.
 *
 * References to AA page numbers are to The Astronomical Almanac, 1986
 * published by the U.S. Government Printing Office.
 *
 * The program uses as a default the PLAN404 approximations to DE404
 * for planetary positions.
 *
 * Warning! Your atan2() function may not work the same as the one
 * assumed by this program.
 * atan2(x,y) computes atan(y/x), result between 0 and 2pi.
 *
 * S. L. Moshier, November, 1987
 */



/* Conversion factors between degrees and radians */
double DTR = 1.7453292519943295769e-2;
double RTD = 5.7295779513082320877e1;
double RTS = 2.0626480624709635516e5; /* arc seconds per radian */
double STR = 4.8481368110953599359e-6; /* radians per arc second */
double PI = 3.14159265358979323846;
extern double PI;

/* Standard epochs.  Note Julian epochs (J) are measured in
 * years of 365.25 days.
 */
double J2000 = 2451545.0;	/* 2000 January 1.5 */
double B1950 = 2433282.423;	/* 1950 January 0.923 Besselian epoch */
double J1900 = 2415020.0;	/* 1900 January 0, 12h UT */

/* Data structures containing orbital elements of
 * objects that orbit the sun.  See kep.h for the definition.
 */
#include "kep.h"

#if __BORLANDC__ || __STDC__
#include <stdlib.h>
#endif

#ifdef _MSC_VER
#if _MSC_VER >= 1000
#include <stdlib.h>
#include <string.h>
#endif
#endif

/* approximate motion of right ascension and declination
 * of object, in radians per day
 */
double FAR dradt;
double FAR ddecdt;

/* Space for star description read from a disc file.
 */
struct star fstar;

/* Space for orbit read from a disc file. Entering 99 for the
 * planet number yields a prompt for a file name containg ASCII
 * strings specifying the elements.
 */
struct orbit forbit;

/* Orbits for each planet.  The indicated orbital elements are
 * not actually used, since the positions are are now calculated
 * from a formula.  Magnitude and semidiameter are still used.
 */
 /* Programs to compute perturbations. */
extern struct plantbl mer404, ven404, ear404, mar404;
extern struct plantbl jup404, sat404, ura404, nep404, plu404;

struct orbit mercury = {
"Mercury        ",
2446800.5, /* January 5.0, 1987 */
7.0048,
48.177,
29.074,
0.387098,
4.09236,
0.205628,
198.7199,
2446800.5,
-0.42,
3.36,
&mer404,
0.0,
0.0,
0.0
};

struct orbit venus = {
"Venus          ",
2446800.5,
3.3946,
76.561,
54.889,
0.723329,
1.60214,
0.006757,
9.0369,
2446800.5,
/* Note the calculated apparent visual magnitude for Venus
 * is not very accurate.
 */
-4.40,
8.34,
&ven404,
0.0,
0.0,
0.0
};

/* Fixed numerical values will be used for earth if read in from a file
 * named earth.orb.  See kfiles.c, kep.h.
 */
struct orbit earth = {
"Earth          ",
2446800.5,
0.0,
0.0,
102.884,
0.999999,
0.985611,
0.016713,
1.1791,
2446800.5,
-3.86,
0.0,
&ear404,
0.0,
0.0,
0.0
};
extern struct orbit earth;

struct orbit mars = {
"Mars           ",
2446800.5,
1.8498,
49.457,
286.343,
1.523710,
0.524023,
0.093472,
53.1893,
2446800.5,
-1.52,
4.68,
&mar404,
0.0,
0.0,
0.0
};

struct orbit jupiter = {
"Jupiter        ",
2446800.5,
1.3051,
100.358,
275.129,
5.20265,
0.0830948,
0.048100,
344.5086,
2446800.5,
-9.40,
98.44,
&jup404,
0.0,
0.0,
0.0
};

struct orbit saturn = {
"Saturn         ",
2446800.5,
2.4858,
113.555,
337.969,
9.54050,
0.0334510,
0.052786,
159.6327,
2446800.5,
-8.88,
82.73,
&sat404,
0.0,
0.0,
0.0
};

struct orbit uranus = {
"Uranus         ",
2446800.5,
0.7738,
73.994,
98.746,
19.2233,
0.0116943,
0.045682,
84.8516,
2446800.5,
-7.19,
35.02,
&ura404,
0.0,
0.0,
0.0
};

struct orbit neptune = {
"Neptune        ",
2446800.5,
1.7697,
131.677,
250.623,
30.1631,
0.00594978,
0.009019,
254.2568,
2446800.5,
-6.87,
33.50,
&nep404,
0.0,
0.0,
0.0
};

struct orbit pluto = {
"Pluto          ",
2446640.5,
17.1346,
110.204,
114.21,
39.4633,
0.00397570,
0.248662,
355.0554,
2446640.5,
-1.0,
2.07,
&plu404,
0.0,
0.0,
0.0
};

/*
int otest(), ctest();
*/
struct orbit test = {
"Test orbit     ",
2446800.5,
1.8498,
49.457,
286.343,
1.523710,
0.524023,
0.093472,
53.1893,
2446800.5,
-1.52,
4.68,
0,
0.0,
0.0,
0.0
};


/* coordinates of object
 */
int objnum = 0;	/* I.D. number of object */
double robject[3] = {0.0}; /* position */
/* ecliptic polar coordinates:
 * longitude, latitude in radians
 * radius in au
 */
double FAR obpolar[3];

/* coordinates of Earth
 */
/* Heliocentric rectangular equatorial position
 * of the earth at time TDT re equinox J2000
 */
double FAR rearth[3];
/* Corresponding polar coordinates of earth:
 * longitude and latitude in radians, radius in au
 */
double FAR eapolar[3];

/* Julian date of ephemeris
 */
double JD;
double TDT;
double UT;

/* flag = 0 if TDT assumed = UT,
 *      = 1 if input time is TDT,
 *      = 2 if input time is UT.
 */
int jdflag = 0;

/* correction vector, saved for display  */
double dp[3];

/* display formats for printf()
 */
extern char *intfmt, *dblfmt;

/* display enable flag
 */
int prtflg = 1;

/* Tabulation parameters
 */
static double djd = 1.0;
static int ntab = 1;

struct orbit *elobject;	/* pointer to orbital elements of object */

/* Main program starts here.
 */
int main()
{
int i;

double zgetdate(), gethms();

kinit();

loop:

prtflg = 1;
printf( "Enter starting date of tabulation\n" );
JD = zgetdate(); /* date */
JD += gethms();	/* time of day */
update(); /* find UT and ET */
printf( "Julian day %.7f\n", JD );

getnum( "Enter interval between tabulations in days", &djd, dblfmt );
getnum( "Number of tabulations to display", &ntab, intfmt );
if( ntab <= 0 )
	ntab = 1;

loop1:
getnum( "Planet number 0-9 or 88 to read star, 99 to read orbit",
	&objnum, intfmt );

switch(objnum)
	{
	case -1: exit(0);
	case 0: elobject = 0;
		printf( "\n                   The Sun\n" );
		break;
	case 1: elobject = &mercury; break;
	case 2: elobject = &venus; break;
	case 3: elobject = 0;
		printf( "\n                   The Moon\n" );
		break;
	case 4: elobject = &mars; break;
	case 5: elobject = &jupiter; break;
	case 6: elobject = &saturn; break;
	case 7: elobject = &uranus; break;
	case 8: elobject = &neptune; break;
	case 9: elobject = &pluto; break;
	case 10: elobject = &test; break;
	case 88:
	        elobject = (struct orbit *)&fstar;
		i = getstar( (struct star *) elobject );
		if( i == 1 )
			goto loop1;
		if( i == 0 )
			break;
		goto operr;
	case 99:
		elobject = &forbit;
		i = getorbit( elobject );
		if( i == 1 )
			goto loop1;
		if( i == 0 )
			break;
	default:
operr:		printf( "Operator error.\n" );
		goto loop;
	}

if( elobject == (struct orbit *)&fstar )
	showcname( &elobject->obname[0] );
else if( elobject )
	printf( "\n                  %s\n", &elobject->obname[0] );

for( i=0; i<ntab; i++ )
	{
/* print Julian date */
	printf( "\nJD %.2f,  ", JD );
	update();

/* Always calculate heliocentric position of the earth */
	kepler( TDT, &earth, rearth, eapolar );

	switch( objnum )
		{
		case 0: dosun(); iter_trnsit( dosun ); break;
		case 3: domoon(); iter_trnsit( domoon ); break;
		case 88: dostar(); iter_trnsit( dostar ); break;
		default: doplanet(); iter_trnsit( doplanet ); break;
		}
	printf( "\n" );
	JD += djd;
	}
goto loop;

#ifdef _MSC_VER
return 0;
#endif
}

